### PR TITLE
RECORDS: can_delete

### PR DESCRIPTION
**Pull request by BadrAly**
NEW: Implement global can_delete

Signed-off-by: Aly Badr <aly.badr@rero.ch>
Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

To Test:

get the PR branch
./run_tests.sh

- Nothing really to test because frontend implementation is planned in Taiga# 775 
- Interface, try to delete any document having items, you should get exception "IlsRecordError.NotDeleted"

### CODE REVIEW

File: rero_ils/modules/documents/api.py [hunk 1]
```diff
@@ -67,3 +68,33 @@ def can_edit(self):
         """Return a boolean for can_edit resource."""
         # TODO: Make this condition on data
         return 'ebook' != self.get('type')
+
+    def get_item_pids(self):
```
Comment (jma) [inline rero_ils/modules/documents/api.py]: get_number_of_items instead

File: tests/ui/patrons/test_patrons_api.py [hunk 2]
```diff
@@ -32,10 +32,10 @@
     patron_id_fetcher
 
 
-def test_patron_create(base_app, roles, user_librarian_data_tmp,
```
Comment (jma) [inline tests/ui/patrons/test_patrons_api.py]: tests are missing for the serializer

File: rero_ils/config.py [hunk 3]
```diff
@@ -303,16 +303,19 @@ def _(x):
         indexer_class=IlsRecordIndexer,
         record_serializers={
             'application/json': (
+                'invenio_records_rest.serializers' ':json_v1_response'
+            ),
+            'application/rero+json': (
```
Comment (jma) [inline rero_ils/config.py]: rero+json can be replaced by can-delete+json

File: rero_ils/modules/item_types/api.py [hunk 4]
```diff
@@ -111,3 +111,19 @@ def exist_name_and_organisation_pid(cls, name, organisation_pid):
             return result.pop(0)
         else:
             return None
+
+    def get_item_pids(self):
+        """Get item pids."""
+        from ..items.api import ItemsSearch
+        results = ItemsSearch().filter(
+            'term', item_type__pid=self.pid).source(['pid']).count()
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/item_types/api.py]: nice catch, done

File: rero_ils/modules/item_types/api.py [hunk 5]
```diff
@@ -111,3 +111,19 @@ def exist_name_and_organisation_pid(cls, name, organisation_pid):
             return result.pop(0)
         else:
             return None
+
+    def get_item_pids(self):
+        """Get item pids."""
+        from ..items.api import ItemsSearch
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/item_types/api.py]: done, whenever possible

File: rero_ils/modules/item_types/api.py [hunk 6]
```diff
@@ -111,3 +111,19 @@ def exist_name_and_organisation_pid(cls, name, organisation_pid):
...
+            'term', item_type__pid=self.pid).source(['pid']).count()
+        return results
+
+    def get_links_to_me(self):
+        """Get number of links."""
+        links = {}
+        items = self.get_item_pids()
+        if items:
+            links['items'] = items
+        # TODO: search for links with circ_policies
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/item_types/api.py]: cases when item type found in `cipo.settings`

File: rero_ils/modules/patron_types/api.py [hunk 7]
```diff
@@ -94,3 +94,18 @@ def exist_name_and_organisation_pid(cls, name, organisation_pid):
...
+        from ..patrons.api import PatronsSearch
+        results = PatronsSearch().filter(
+            'term', patron_type__pid=self.pid).source(['pid']).count()
+        return results
+
+    def get_links_to_me(self):
+        """Get number of links."""
+        links = {}
+        patrons = self.get_patron_pids()
+        if patrons:
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/patron_types/api.py]: done

File: rero_ils/modules/libraries/api.py [hunk 8]
```diff
@@ -239,3 +239,17 @@ def in_working_days(self, count, date=datetime.now()):
...
+    def get_location_pids(self):
+        """Get library location pids."""
+        results = LocationsSearch().filter(
+            'term', library__pid=self.pid).source(['pid']).count()
+        return results
+
+    def get_links_to_me(self):
+        """Get number of links."""
+        links = {}
+        locations = self.get_location_pids()
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/libraries/api.py]: added users with role `librarian`

File: rero_ils/modules/documents/api.py (hunk 1 above)
Comment (BadrAly) [reply rero_ils/modules/documents/api.py]: done

File: rero_ils/modules/serializers.py [hunk 9]
```diff
@@ -46,6 +46,11 @@ def serialize(self, pid, record, links_factory=None, **kwargs):
         return super(
             ReroIlsSerializer, self).serialize(
                 pid, record, links_factory, **kwargs)
+        rec = self.transform_record(
+            pid, record, links_factory, **kwargs
+        ).get('metadata', {})
+        rec['links_to_me'] = record.get_links_to_me()
+        return json.dumps(rec, **self._format_args())
```
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/serializers.py]: done

File: rero_ils/config.py (hunk 3 above)
Comment (BadrAly) [reply rero_ils/config.py]: done, there is possibly something to adapt on the admin views for this as well (?)

File: rero_ils/modules/serializers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (BadrAly) [reply rero_ils/modules/serializers.py]:
```cannot_delete = {
        "links": {
            "items": 5
        },
        "others" : {
            "is_default": true,
            "has_settings": 5
        }
}

File: tests/ui/patrons/test_patrons_api.py (hunk 2 above)
Comment (BadrAly) [reply tests/ui/patrons/test_patrons_api.py]: done

**Pull request merged**